### PR TITLE
Cognito provider

### PR DESCRIPTION
**Pull request by bretambrose**
* Cognito credentials provider binding
* STS credentials provider binding

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/crt/auth/Credentials.h [hunk 1]
```diff
@@ -335,6 +335,118 @@ namespace Aws
...
+                /**
+                 * How long sourced credentials should remain valid for, in seconds.  900 is the minimum allowed value.
+                 */
+                uint16_t DurationSeconds;
+
+                /**
+                 * Connection bootstrap to use to create the http connection required to
+                 * query credentials from the STS provider
+                 *
+                 * Note: If null, then the default ClientBootstrap is used
```
Comment (TwistedTwigleg) [inline include/aws/crt/auth/Credentials.h:434]: Super minor nitpick: `null` -> `nullptr`

File: include/aws/crt/auth/Credentials.h [hunk 2]
```diff
@@ -448,6 +560,20 @@ namespace Aws
                     const CredentialsProviderDelegateConfig &config,
                     Allocator *allocator = ApiAllocator());
 
+                /**
+                 * Creates a provider that sources credentials from the Cognito Identity service
```
Comment (TwistedTwigleg) [inline include/aws/crt/auth/Credentials.h:564]: Super minor nitpick: Should `Cognito Identity service` be `Cognito credentials provider`? Or are they separate things?

File: source/auth/Credentials.cpp [hunk 3]
```diff
@@ -372,6 +372,98 @@ namespace Aws
...
+            CredentialsProviderSTSConfig::CredentialsProviderSTSConfig() : Bootstrap(nullptr) {}
+
+            std::shared_ptr<ICredentialsProvider> CredentialsProvider::CreateCredentialsProviderSTS(
+                const CredentialsProviderSTSConfig &config,
+                Allocator *allocator)
+            {
+                struct aws_credentials_provider_sts_options raw_config;
+                AWS_ZERO_STRUCT(raw_config);
+
+                raw_config.creds_provider = config.Provider->GetUnderlyingHandle();
```
Comment (TwistedTwigleg) [inline source/auth/Credentials.cpp:444]: Question: Do we need to check if a provider is defined?

**Review by TwistedTwigleg — APPROVED** [review]
Looks good! There was a few minor nitpicks and a question, but they are minor and it looks good to me 👍

File: include/aws/crt/auth/Credentials.h (hunk 1 above)
Comment (bretambrose) [reply include/aws/crt/auth/Credentials.h:434]: While nullptr is the C++11 syntax, colloquially I always hear "null" (is-null, is-not-null, etc...)  It feels really weird and awkward to me the other way.

File: include/aws/crt/auth/Credentials.h (hunk 2 above)
Comment (bretambrose) [reply include/aws/crt/auth/Credentials.h:564]: The Cognito-Identity is the web service that the provider sources from, so it should be good as written.

File: source/auth/Credentials.cpp (hunk 3 above)
Comment (bretambrose) [reply source/auth/Credentials.cpp:444]: Yeah, that seems like a good check.

File: include/aws/crt/auth/Credentials.h (hunk 1 above)
Comment (TwistedTwigleg) [reply include/aws/crt/auth/Credentials.h:434]: That is fair. In that case, then I'd leave it as `null` 👍

**Pull request merged**